### ISSUE TITLE
Option to run serial even when OpenCL is available

### ISSUE THREAD

**Issue by bentsherman** [SystemsGenetics/ACE#22]
Since ACE automatically connects to the first OpenCL device that is available, there is no way to revert to the serial implementation except to disable the analytic's OpenCL capabilities in code and rebuild the application (or something similar). I find myself having to do this a lot during development in order to test each implementation, but I think it might also come up after development for doing things like performance comparisons.

I think it's useful that ACE can connect to device 0 automatically, but I think it should be complemented with the ability to "clear" the current OpenCL device. Maybe with a checkbox in the OpenCL device dialog? Or perhaps a button that just clears the device.

**Comment by 4ctrl-alt-del** [comment]
This is on my list of things to do :)

**Comment by bentsherman** [comment]
I may work on this soon if I have time, since it will really help me do performance comparisons. For anyone who is interested (including future me), here is what I'm thinking:

- For ACE gui, just add a "Clear" button to the "Set OpenCL Device" menu, maybe something to show the current device if there isn't something for that already.
- For ACE console, implement a `run-serial` command which mimics the `run` command and forces the analytic to use serial code. In the future we could also add commands `run-opencl` and `run-cuda` to extend this behavior to those capabilities.

**Comment by bentsherman** [comment]
The `force-serial` branch just has the `run-serial` command added for ACE console. For the GUI, now I'm thinking it would make more sense to add a "Force Serial" option under the settings menu. That will be much easier to extend if we add more capabilities in the future.

**Comment by bentsherman** [comment]
Force-updated the `force-serial` branch to include MPI fixes from `master`. I was trying to set this up so that I could go ahead and use it for performance testing, but it seems like adding the `_forcedCapability` member to `EAbstractAnalytic` leads to clobbering with other class members. That is, depending on where I placed `_forcedCapability` in the class definition or where I call `forceCapability()` in `run()`, I'll get argument errors that don't make sense, segfaults, or worse. The most I can say from debugging is that maybe it's because EAbstractAnalytic's default constructor is never actually called, so maybe it's members aren't being initialized properly? The `_forcedCapability` member definitely doesn't get initialized to 0 even though I set it in the class definition.

**Comment by 4ctrl-alt-del** [comment]
In C++ if no constructor is defined the default constructor with no arguments is implemented. Try adding an initializer to that variable? bool _forcedCapacility {false};

**Comment by 4ctrl-alt-del** [comment]
Also try cleaning the project, run qmake, and rebuild.

**Comment by bentsherman** [comment]
Yeah, so I do initialize it to 0 (it's a `quint32`), but during debugging I found that it was some garbage number, as if it hadn't been initialized or had been clobbered by something else. So I figured either the constructor wasn't being called or _forcedCapability is being clobbered somehow when the analytic is initialized and loaded up with arguments. I don't know why that would happen, it's really weird what's going on. I think I've tried rebuilding, but I'll try it again later just to be sure.
